internal/providers/registry.go

```diff
@@ -1,52 +1,318 @@
 package providers
 
 import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
 	"sync"
+	"time"
 
+	"github.com/filecoin-project/storetheindex/config"
+	"github.com/filecoin-project/storetheindex/internal/providers/discovery"
+	"github.com/filecoin-project/storetheindex/internal/providers/policy"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	logging "github.com/ipfs/go-log/v2"
 	"github.com/libp2p/go-libp2p-core/peer"
-	ma "github.com/multiformats/go-multiaddr"
 )
 
+var log = logging.Logger("providers")
+
 type Registry struct {
-	rwmutex   sync.RWMutex
+	actions   chan func()
 	providers map[peer.ID]*ProviderInfo
+	policy    *policy.Policy
+
+	discovery    discovery.Discovery
+	discoTimes   map[string]time.Time
+	closeOnce    sync.Once
+	closed       chan struct{}
+	pendingDisco int
+
+	dstore datastore.Datastore
 }
 
+// ProviderInfo is an immutable data sturcture that holds information about a
+// provider.  A ProviderInfo instance is never modified, but rather a new one
+// is created to update its contents.  This means existing references remain
+// valid.
 type ProviderInfo struct {
-	Addresses []ma.Multiaddr
+	AddrInfo      peer.AddrInfo
+	FcAddress     string
```

Comment:
better name?

Reply:
Is there a more generic name for the address that we identify a miner with to lotus (e.g., `t01000`)?  Should I call this `MinerAddress`, `OnChainID`, `DiscoveryID`?

Reply:
If this were an IPFS provider to discover, then this would be a Multiaddr and would usually be the same one of the discovered addresses (which might include multiple addresses). So, I think a better name is DiscoverAddr. What is missing is where/how the discovery was done.

Reply:
Renamed `FcAddress` to `DiscoveryAddr`